### PR TITLE
fix(core): Store the currently selected ICU in `LView`

### PR DESCRIPTION
**Pull request by mhevery**
This PR builds on:
- [x] [refactor(core): add `debug` for i18n #38154](https://github.com/angular/angular/pull/38154) 
- [ ] [refactor(core): add debug ranges to `LViewDebug` #38359](https://github.com/angular/angular/pull/38359)

---

The currently selected ICU was incorrectly being stored it `TNode`
rather than in `LView`.

Remove: `TIcuContainerNode.activeCaseIndex`
Add: `LView[TIcu.currentCaseIndex]`

## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?

- [ ] Yes
- [ ] No

### CODE REVIEW

File: packages/core/src/render3/i18n.ts [hunk 1]
```diff
@@ -932,68 +950,96 @@ function readUpdateOpCodes(
...
+ * @param icuIndex Index into `tIcus` to process.
+ * @param bindingsStartIndex Location of the first `ɵɵi18nApply`
+ * @param lView Current `LView`
+ * @param changeMask Each bit corresponds to a `ɵɵi18nExp` (Counting backwards from
+ *     `bindingsStartIndex`)
+ */
+function applyIcuUpdateCase(
+    tView: TView, tIcus: TIcu[], icuIndex: number, bindingsStartIndex: number, lView: LView,
+    caseCreated: boolean) {
+  const tIcu = tIcus[icuIndex];
```
Comment (AndrewKushnir) [inline packages/core/src/render3/i18n.ts:967]: nit: may be we can add an assert that `icuIndex` is is in the right range (`[0, tIcus.length - 1]`)?

File: packages/core/src/render3/i18n.ts [hunk 2]
```diff
@@ -932,68 +950,96 @@ function readUpdateOpCodes(
...
+ *
+ * @param tView Current `TView`
+ * @param tIcus ICUs active at this location.
+ * @param icuIndex Index into `tIcus` to process.
+ * @param lView Current `LView`
+ * @returns true if a new case was created (needed so that the update executes regardless of the
+ *     bitmask)
+ */
+function applyIcuSwitchCaseRemove(tView: TView, tIcus: TIcu[], icuIndex: number, lView: LView) {
+  const tIcu = tIcus![icuIndex];
```
Comment (AndrewKushnir) [inline packages/core/src/render3/i18n.ts:1022]:
```suggestion
  const tIcu = tIcus[icuIndex];
```
It should not be needed based on the `tIcus` type.

File: packages/core/src/render3/i18n.ts [hunk 3]
```diff
@@ -932,68 +950,96 @@ function readUpdateOpCodes(
...
+function applyIcuSwitchCase(
+    tView: TView, tIcus: TIcu[], icuIndex: number, lView: LView, value: string): boolean {
+  applyIcuSwitchCaseRemove(tView, tIcus, icuIndex, lView);
+
+  // Rebuild a new case for this ICU
   let caseCreated = false;
-  // If there is an active case, delete the old nodes
-  if (icuTNode.activeCaseIndex !== null) {
-    const removeCodes = tIcu.remove[icuTNode.activeCaseIndex];
+  const tIcu = tIcus[icuIndex];
```
Comment (AndrewKushnir) [inline packages/core/src/render3/i18n.ts:996]: It looks like we can calculate this at the very beginning of the function and pass `tIcu` to the `applyIcuSwitchCaseRemove` function call (instead of passing `tIcus` and `icuIndex` and calculating the current ICU there).

File: packages/core/src/render3/i18n.ts [hunk 4]
```diff
@@ -1239,6 +1286,9 @@ function icuStart(
   const tIcu: TIcu = {
     type: icuExpression.type,
     vars,
+    currentCaseIndex: HEADER_OFFSET +
+        expandoStartIndex  // expandoStartIndex does not include the header so add it.
+        + 1,               // The first item stored is the `<!--ICU #-->` anchor so skip it.
```
Comment (AndrewKushnir) [inline packages/core/src/render3/i18n.ts:1291]: Since this is a new slot in an `LView` that we are allocating, should we also increase the space in expando block (i.e. call `allocExpando` with the `i18nVarsCount + 1`)?

File: packages/core/src/render3/instructions/shared.ts [hunk 5]
```diff
@@ -240,8 +240,7 @@ export function getOrCreateTNode(
     // For performance reasons it is important that the tNode retains the same shape during runtime.
     // (To make sure that all of the code is monomorphic.) For this reason we seal the object to
     // prevent class transitions.
-    // FIXME(misko): re-enable this once i18n code is compliant with this.
-    // Object.seal(tNode);
+    Object.seal(tNode);
```
Comment (AndrewKushnir) [inline packages/core/src/render3/instructions/shared.ts:243]: I feel like we may need to extract this change in it's own PR to minimize risks in this PR.

File: packages/core/src/render3/interfaces/i18n.ts [hunk 6]
```diff
@@ -346,6 +346,14 @@ export interface TIcu {
    */
   vars: number[];
 
+  /**
+   * Currently selected ICU case pointer.
+   *
+   * `lView[currentCaseIndex]` stores the currently selected case. This is needed to know how to
+   * clean up the current case when transitioning no the new case.
+   */
+  currentCaseIndex: number;
```
Comment (AndrewKushnir) [inline packages/core/src/render3/interfaces/i18n.ts:355]: I was initially confused that we still keep the current case index in T-structure (as the name of this field suggests), but this is a pointer to an LView, so everything is great. I was just thinking if we can give this field slightly different name to avoid this confusion in the future. However finding a good name for it is tricky, since it's "an index to an index" :) May be something like `currentCaseLViewIndex`?

**Review by AndrewKushnir — COMMENTED** [review]
Thanks for the fix @mhevery 👍  I've just left a few comments.

File: packages/core/src/render3/instructions/shared.ts (hunk 5 above)
Comment (mhevery) [reply packages/core/src/render3/instructions/shared.ts:243]: Move the `seal` into `createTNodeAtIndex`

Comment (mhevery) [reply packages/core/src/render3/instructions/shared.ts:243]: Maybe add `seal` to other data structures as well.

File: packages/core/src/render3/i18n.ts (hunk 1 above)
Comment (mhevery) [reply packages/core/src/render3/i18n.ts:967]: added.

File: packages/core/src/render3/i18n.ts (hunk 3 above)
Comment (mhevery) [reply packages/core/src/render3/i18n.ts:996]: Yes, we could, and I originally had it that way, but it does not work for all of the apply methods, so I think for consistency I would prefer to pass in `tIcus: TIcu[], icuIndex: number` to all of these methods. Is that OK with you?

File: packages/core/src/render3/i18n.ts (hunk 2 above)
Comment (mhevery) [reply packages/core/src/render3/i18n.ts:1022]: fixed

File: packages/core/src/render3/i18n.ts (hunk 4 above)
Comment (mhevery) [reply packages/core/src/render3/i18n.ts:1291]: Yes, we need to do that. That is done later. See: 
```
vars: 1,  // allocate space for `TIcu.currentCaseIndex`
```

File: packages/core/src/render3/instructions/shared.ts (hunk 5 above)
Comment (mhevery) [reply packages/core/src/render3/instructions/shared.ts:243]: fixed

File: packages/core/src/render3/interfaces/i18n.ts (hunk 6 above)
Comment (mhevery) [reply packages/core/src/render3/interfaces/i18n.ts:355]: done.

**Comment by mhevery** [comment]
[presubmit](http://test/OCL:325328186:BASE:325328202:1596754395777:fd5d67c9)

File: packages/core/src/render3/i18n.ts [hunk 7]
```diff
@@ -932,68 +950,98 @@ function readUpdateOpCodes(
...
+ * @param tView Current `TView`
+ * @param tIcus If ICUs present than this contain them.
+ * @param icuIndex Index into `tIcus` to process.
+ * @param bindingsStartIndex Location of the first `ɵɵi18nApply`
+ * @param lView Current `LView`
+ * @param changeMask Each bit corresponds to a `ɵɵi18nExp` (Counting backwards from
+ *     `bindingsStartIndex`)
+ */
+function applyIcuUpdateCase(
+    tView: TView, tIcus: TIcu[], icuIndex: number, bindingsStartIndex: number, lView: LView,
```
Comment (AndrewKushnir) [inline packages/core/src/render3/i18n.ts:965]: nit: I'd propose to rename `icuIndex` -> `tIcuIndex` to establish a relation with the `tIcus` argument.
```suggestion
    tView: TView, tIcus: TIcu[], tIcuIndex: number, bindingsStartIndex: number, lView: LView,
```

File: packages/core/src/render3/i18n.ts [hunk 8]
```diff
@@ -932,68 +950,98 @@ function readUpdateOpCodes(
...
-  const icuTNode = getTNode(tView, nodeIndex) as TIcuContainerNode;
-  if (icuTNode.activeCaseIndex !== null) {
-    readUpdateOpCodes(
-        tIcu.update[icuTNode.activeCaseIndex], tIcus, bindingsStartIndex, changeMask, tView, lView,
-        caseCreated);
+/**
+ * Apply OpCodes associated with updating an existing ICU.
+ *
+ * @param tView Current `TView`
+ * @param tIcus If ICUs present than this contain them.
```
Comment (AndrewKushnir) [inline packages/core/src/render3/i18n.ts:957]: I believe in case of `applyIcuUpdateCase`, ICUs are always present.
```suggestion
 * @param tIcus ICUs active at this location.
```

File: packages/core/src/render3/i18n.ts [hunk 9]
```diff
@@ -932,68 +950,98 @@ function readUpdateOpCodes(
...
+ * @param tView Current `TView`
+ * @param tIcus ICUs active at this location.
+ * @param icuIndex Index into `tIcus` to process.
+ * @param lView Current `LView`
+ * @param value Value of the case to update to.
+ * @returns true if a new case was created (needed so that the update executes regardless of the
+ *     bitmask)
+ */
+function applyIcuSwitchCase(
+    tView: TView, tIcus: TIcu[], icuIndex: number, lView: LView, value: string): boolean {
```
Comment (AndrewKushnir) [inline packages/core/src/render3/i18n.ts:993]:
```suggestion
    tView: TView, tIcus: TIcu[], tIcuIndex: number, lView: LView, value: string): boolean {
```

File: packages/core/src/render3/i18n.ts (hunk 3 above)
Comment (AndrewKushnir) [reply packages/core/src/render3/i18n.ts:998]: Sounds good 👍

File: packages/core/src/render3/i18n.ts [hunk 10]
```diff
@@ -932,68 +950,98 @@ function readUpdateOpCodes(
...
+  applyIcuSwitchCaseRemove(tView, tIcus, icuIndex, lView);
+
+  // Rebuild a new case for this ICU
   let caseCreated = false;
-  // If there is an active case, delete the old nodes
-  if (icuTNode.activeCaseIndex !== null) {
-    const removeCodes = tIcu.remove[icuTNode.activeCaseIndex];
+  const tIcu = tIcus[icuIndex];
```
Comment (AndrewKushnir) [inline packages/core/src/render3/i18n.ts:998]: May be add the checks that the index is in the range here as well?
```suggestion
  ngDevMode && assertIndexInRange(tIcus, icuIndex);
  const tIcu = tIcus[icuIndex];
```

File: packages/core/src/render3/i18n.ts [hunk 11]
```diff
@@ -932,68 +950,98 @@ function readUpdateOpCodes(
...
+ *
+ * @param tView Current `TView`
+ * @param tIcus ICUs active at this location.
+ * @param icuIndex Index into `tIcus` to process.
+ * @param lView Current `LView`
+ * @returns true if a new case was created (needed so that the update executes regardless of the
+ *     bitmask)
+ */
+function applyIcuSwitchCaseRemove(tView: TView, tIcus: TIcu[], icuIndex: number, lView: LView) {
+  const tIcu = tIcus[icuIndex];
```
Comment (AndrewKushnir) [inline packages/core/src/render3/i18n.ts:1024]: May be add the checks that the index is in the range here as well?
```suggestion
  ngDevMode && assertIndexInRange(tIcus, icuIndex);
  const tIcu = tIcus[icuIndex];
```

**Review by AndrewKushnir — APPROVED** [review]
Looks great (just left a few minor comments) 👍

File: packages/core/src/render3/i18n.ts (hunk 8 above)
Comment (mhevery) [reply packages/core/src/render3/i18n.ts:957]: fixed

File: packages/core/src/render3/i18n.ts (hunk 7 above)
Comment (mhevery) [reply packages/core/src/render3/i18n.ts:965]: done

File: packages/core/src/render3/i18n.ts (hunk 9 above)
Comment (mhevery) [reply packages/core/src/render3/i18n.ts:993]: done

File: packages/core/src/render3/i18n.ts (hunk 10 above)
Comment (mhevery) [reply packages/core/src/render3/i18n.ts:998]: done

File: packages/core/src/render3/i18n.ts (hunk 11 above)
Comment (mhevery) [reply packages/core/src/render3/i18n.ts:1024]: done

**Review by atscott — APPROVED** [review]
reviewed-for: size-tracking

**Review by AndrewKushnir — APPROVED** [review]
Reviewed-for: size-tracking
